### PR TITLE
[unitaryhack] Add towncrier for managing the changelog

### PR DESCRIPTION
**Pull request by trentfridey**
**Description**
This sets up the towncrier utility to manage the changelog

- Add suitable towncrier configuration to the respository.
- Update the release documentation with instructions for using towncrier to update the change log.
- Update the development documentation to include the requirement to add towncrier entries for each PR.
- Update the GitHub PR template to mention the requirement to add a towncrier entry.
- Add a check to the GitHub action tests workflow that checks that a towncrier entry was added.

**Related issues or PRs**
#1894

**Changelog**
See doc/changes/1927.doc

### CODE REVIEW

**Comment by hodgestar** [comment]
@trentfridey Thank you for looking at this. It looks good to me, but I'd like @Ericgig and @AGaliciaMartinez to read and approve too since they're also developing very actively.

I see this is still marked as a work in progress -- are there changes you still want to make from your side?

I guess we also need to add fragments for all the PRs already merged for milestone 4.7.1. If you want to give that a try that would be awesome, but otherwise I'm happy to do it next week when I have a bit more time again.

**Review by AGaliciaMartinez — CHANGES_REQUESTED** [review]
Thank you for the contribution! I added some suggestions but my biggest concern is regarding how things are explained in the documentaiton. I think we should try to not force the user to install towncrier _and_ use it from the command line. This is fine with me but I think it will be a barrier for new contributors. Since it is possible to manually create a file with the changelog, we should favor the manual method as it does not require any additional setup for the user. For instance, this is what pytest does in their repository (https://github.com/pytest-dev/pytest/blob/main/CONTRIBUTING.rst). I think an approach similar to the one followed by pytest would suit us best.

**Comment by Ericgig** [comment]
Is there a way to set towncrier to not require the PR number?
Otherwise we need to open a PR without the changelog, which will fail the towncrier test, to modify it to add the change log...
It would a lot better if we could add it when developing or have a towncrier read it from the Changelog section of PR description.

**Comment by trentfridey** [comment]
> Is there a way to set towncrier to not require the PR number? Otherwise we need to open a PR without the changelog, which will fail the towncrier test, to modify it to add the change log... It would a lot better if we could add it when developing or have a towncrier read it from the Changelog section of PR description.

Yes, I think we could use the issue number that the change is for, if this workflow makes sense for QuTiP (I don't know, I'm a first-time contributor). This is what [pytest does](https://github.com/pytest-dev/pytest/blob/main/CONTRIBUTING.rst#preparing-pull-requests).

File: doc/development/contributing.rst
Comment (trentfridey) [reply doc/development/contributing.rst:148]: Yes I can change this once we have figured out what we want to do about the PR number being required as @Ericgig pointed out

**Comment by AGaliciaMartinez** [comment]
Unfortunately, not all the PR we open are related to an existing issue.

Could it be possible to include the creation of the changelog as part of a github workflow? If we are able to read the PR number and the Changelog section of the PR description in a workflow, we may be able to create the changelog file during the merge step to later be used with towncrier. This would also mean that new contributors would only require to write the Changelog section of the PR description (which they already have to do), instead of having to create a new file.

**Comment by hodgestar** [comment]
We currently link to the PR in the changelog text, and that that seems a good thing to keep, so we already have to write or update the text after the PR is created.

I'm a bit against pulling information from the PR description. My observation is that the descriptions (even my own) are not very good, because they're written when the PR is just opened, before all of the improvements that happen during review.

@trentfridey Does the PR number have to appear in the filename or confer any special advantages? Or can one just use a meaningful unique name for the file? E.g. `fix-20220614-add-towncrier`?

**Review by hodgestar — APPROVED** [review]

**Comment by hodgestar** [comment]
@Ericgig @AGaliciaMartinez I've approved this PR and I'd like to merge it once @trentfridey has fixed the "Verify Towncrier entry added". I actually preferred the situation before @trentfridey removed the requirement for the PR number -- it was simpler and didn't rely on us to remember to write those numbers into the changelog.

For UnitaryHack, I think we have to merge it before the end of tomorrow for it to count, and I'd like to do that since it's a big improvement and now probably we just need to try it and tweak how it is set up if needed.

File: doc/development/contributing.rst
Comment (trentfridey) [reply doc/development/contributing.rst]: addressed in [79b9c1f](https://github.com/qutip/qutip/pull/1927/commits/79b9c1fcec317c5b6d843adb678d4e96a2a0da2d)

**Comment by AGaliciaMartinez** [comment]
I also prefer to have the PR number for reference in the changelog. Otherwise this looks great. Thanks @trentfridey!

**Review by AGaliciaMartinez — APPROVED** [review]

**Comment by AGaliciaMartinez** [comment]
@trentfridey, there seems to be an error in the newly added towncrier github workflow that needs to be fixed before merging. Would you mind taking a look?

**Comment by Ericgig** [comment]
I don't want to approve or block it.
I can't understand the template so I don't know if the generated changelog will be useful for release.
And the tests don't pass...

Personally the issue number as the tag looks good to me. Yes we have many PRs without issues, but it may be a good idea to open them just to tell other what we are working on and it's easier for first time contributor who take a `good first issue`.

**Comment by hodgestar** [comment]
> I don't want to approve or block it.

I don't want to rush it either. I asked the UnitaryHack organizers if we could merge it after the deadline and still have it count for @trentfridey and they said yes!

Let's restore things to the previous "PR name required and no template" state and then we can merge something that we all like without rushing too much.

@trentfridey Thanks for all your work so far and @Ericgig and @AGaliciaMartinez thank you for reviewing with a bit of a deadline (which is not how we usually work).

**Comment by trentfridey** [comment]
Ok, I've reverted to using the PR number for the news fragment file names, and updated the contributing/PR instructions.

I also looked into why the `towncrier-check` test is failing. It looks like towncrier is not finding the `pyproject.toml` config file because it is not running in the root directory. I added a `working-directory: /` specification to the action step to address this, hopefully this should fix it.

File: pyproject.toml
Comment (hodgestar) [inline pyproject.toml:36]:
```suggestion
directory = "feature"
```

File: pyproject.toml
Comment (hodgestar) [inline pyproject.toml:40]:
```suggestion
directory = "bugfix"
```

**Review by hodgestar — APPROVED** [review]
I made two small suggestions to fix typos in the names of change types in the .toml file. I will apply these shortly.

**Comment by hodgestar** [comment]
@AGaliciaMartinez @Ericgig Looks good to me. I will merge later this afternoon if there are no objections.

File: .github/workflows/tests.yml
Comment (hodgestar) [inline .github/workflows/tests.yml:187]: GitHub rejects `working-directory` here as invalid -- it would need to be inside the step (i.e. same level a `run:`) -- but I'm also moderately sure that the default working directory is already the folder that contains `pyproject.toml`, so I'm not sure what is happening here.

**Review by hodgestar — CHANGES_REQUESTED** [review]

File: pyproject.toml
Comment (hodgestar) [inline pyproject.toml:28]:
```suggestion
```
Remove the package configuration -- we are using the directory directly.

**Review by hodgestar — APPROVED** [review]
I fixed the issue, which was firstly that the repo hadn't been checked out at all, and then that GitHub doesn't check out enough of the history of the repository by default in GitHub actions.

**Comment by nathanshammah** [comment]
@trentfridey we cannot find your contact details on unitaryHACK, please sign up https://airtable.com/shrJeycewBFqdot2B to be able to claim your bounty!

**Comment by trentfridey** [comment]
I just signed up. Let me know if you need anything else.

**Pull request merged**
